src/native/managed/cdacreader/src/Contracts/Registry.cs

```diff
@@ -0,0 +1,32 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Diagnostics.DataContractReader.Contracts;
+
+internal sealed class Registry
+{
+    public Dictionary<Type, IContract> _contracts = [];
+    public Target _target;
+
+    public Registry(Target target)
+    {
+        _target = target;
+    }
+
+    public Thread Thread => GetContract<Thread>();
+
+    private T GetContract<T>() where T : IContract
+    {
+        if (_contracts.TryGetValue(typeof(T), out IContract? contractMaybe))
+            return (T)contractMaybe;
+
+        IContract contract = T.Create(_target);
+
+        // Still okay if contract was already registered by someone else
```

Comment:
Why is this okay? Does the `contract` not really matter in this case? The below logic means we could fail to add the `contract` because one already exists, and then return the one we just created. This seems like we could get into a state where an implementation has state and isn't the one being used. I think we need some other invariant in the system if this pattern is considered safe.

Reply:
> I don't think the contracts should really have state other than what is tied to the target itself.

I kind of want contract implementations to be valuetypes that way its evident they have no identity.  Not sure if that's going overboard, though

Reply:
> That way it's evident they have no identity. Not sure if that's going overboard, though

That is the kind of architecture invariant that would work for me. It isn't a compile time enforcement, but does make people ask the right question. The counter to that though then becomes, why do we have a cache if none of the objects have state? Will the constructor have a lot of logic to set itself up?

Reply:
> Will the constructor have a lot of logic to set itself up?

My feeling is, yes. (well not literally the constructor, but the `TryCreate` static virtual method).  In the C++ prototype there was a lot of work upfront to query the target and establish that the necessary data was available.  I think something similar is already evident in the `Thread` contract, too